File: common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/bridge/BridgeRangeCalculator.java
package net.shiroha233.roadweaver.features.path.pathlogic.bridge;

import net.minecraft.core.BlockPos;
import net.shiroha233.roadweaver.config.ConfigService;
import net.shiroha233.roadweaver.config.ModConfig;
import net.shiroha233.roadweaver.core.model.RoadSpan;
import net.shiroha233.roadweaver.core.model.SpanType;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 桥梁范围计算器
 */
public final class BridgeRangeCalculator {
    private BridgeRangeCalculator() {}

    public static record RangeResult(boolean[] isBridge, List<int[]> mergedRanges, boolean[] skipSegments) {}

    private static double dist2d(BlockPos a, BlockPos b) {
        long dx = (long) b.getX() - a.getX();
        long dz = (long) b.getZ() - a.getZ();
        return Math.sqrt((double) dx * dx + (double) dz * dz);
    }

    private static int estimateRangeLengthBlocks(List<BlockPos> middlePositions, int a, int b) {
        if (middlePositions == null || middlePositions.isEmpty()) return 0;
        int n = middlePositions.size();
        int from = Math.max(0, Math.min(n - 1, a));
        int to = Math.max(0, Math.min(n - 1, b));
        if (from > to) {
            int t = from;
            from = to;
            to = t;
        }
        double sum = 0.0;
        for (int i = from; i < to; i++) {
            sum += dist2d(middlePositions.get(i), middlePositions.get(i + 1));
        }
        return (int) Math.round(sum);
    }

    public static RangeResult compute(List<BlockPos> middlePositions, List<RoadSpan> spans) {
        ModConfig cfg = ConfigService.get();
        return compute(middlePositions, spans, cfg, null);
    }

    public static RangeResult compute(List<BlockPos> middlePositions,
                                      List<RoadSpan> spans,
                                      ModConfig cfg,
                                      String dimensionId) {
        int n = middlePositions.size();
        boolean[] isBridge = new boolean[n];
        boolean[] skipSegments = new boolean[n];
        List<int[]> bridgeRanges = new ArrayList<>();

        if (cfg == null) {
            return new RangeResult(isBridge, List.of(), skipSegments);
        }

        boolean bridgeEnabled = (dimensionId == null || dimensionId.isEmpty())
                ? cfg.bridgeEnabled()
                : cfg.bridgeEnabledForDimension(dimensionId);
        if (!bridgeEnabled) {
            return new RangeResult(isBridge, List.of(), skipSegments);
        }

        int minLength = Math.max(1, cfg.bridgeMinLength());
        int mergeGap = Math.max(1, cfg.bridgeMergeGap());
        boolean useBuoysInstead = cfg.bridgeUseBuoysInstead();
        int maxLenBlocks = useBuoysInstead ? 0 : Math.max(0, cfg.bridgeMaxLengthBlocks());
        
        if (spans != null) {
            Map<Long, Integer> indexMap = new HashMap<>();
            for (int i = 0; i < n; i++) indexMap.put(middlePositions.get(i).asLong(), i);
            for (RoadSpan sp : spans) {
                if (sp.type() != SpanType.BRIDGE) continue;
                Integer si = indexMap.get(sp.start().asLong());
                Integer ei = indexMap.get(sp.end().asLong());
                if (si == null || ei == null) continue;
                int a = Math.max(0, Math.min(si, ei));
                int b = Math.min(n - 1, Math.max(si, ei));
                bridgeRanges.add(new int[]{a, b});
            }
        }
        
        if (!bridgeRanges.isEmpty()) {
            bridgeRanges.sort(Comparator.comparingInt(o -> o[0]));
            
            List<int[]> merged = new ArrayList<>();
            int[] cur = bridgeRanges.get(0);
            for (int idx = 1; idx < bridgeRanges.size(); idx++) {
                int[] nxt = bridgeRanges.get(idx);
                if (nxt[0] <= cur[1] + mergeGap) {
                    cur[1] = Math.max(cur[1], nxt[1]);
                } else {
                    merged.add(cur);
                    cur = nxt;
                }
            }
            merged.add(cur);
            
            List<int[]> filtered = new ArrayList<>();
            for (int[] r : merged) {
                int len = r[1] - r[0] + 1;
                if (len >= minLength) {
                    filtered.add(r);
                }
            }
            List<int[]> filteredByMaxLen = new ArrayList<>();
            for (int[] r : filtered) {
                if (maxLenBlocks <= 0) {
                    filteredByMaxLen.add(r);
                    continue;
                }
                int approxLen = estimateRangeLengthBlocks(middlePositions, r[0], r[1]);
                if (approxLen > maxLenBlocks) {
                    int s = Math.max(0, r[0]);
                    int e = Math.min(n - 1, r[1]);
                    for (int k = s; k <= e; k++) {
                        skipSegments[k] = true;
                    }
                } else {
                    filteredByMaxLen.add(r);
                }
            }
            bridgeRanges = filteredByMaxLen;
        }
        
        for (int[] r : bridgeRanges) {
            for (int k = r[0]; k <= r[1]; k++) {
                isBridge[k] = true;
            }
        }
        
        return new RangeResult(isBridge, bridgeRanges, skipSegments);
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/features/path/pathlogic/bridge/BridgeRangeCalculator.kt
package net.shiroha233.roadweaver.features.path.pathlogic.bridge

import net.minecraft.core.BlockPos
import net.shiroha233.roadweaver.config.ConfigService
import net.shiroha233.roadweaver.config.ModConfig
import net.shiroha233.roadweaver.helpers.Records
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round
import kotlin.math.sqrt

object BridgeRangeCalculator {
    data class RangeResult(
        val isBridge: BooleanArray,
        val mergedRanges: List<IntArray>,
        val skipSegments: BooleanArray
    )

    private fun dist2d(a: BlockPos, b: BlockPos): Double {
        val dx = (b.x.toLong() - a.x.toLong())
        val dz = (b.z.toLong() - a.z.toLong())
        return sqrt(dx.toDouble() * dx.toDouble() + dz.toDouble() * dz.toDouble())
    }

    private fun estimateRangeLengthBlocks(middlePositions: List<BlockPos>?, a: Int, b: Int): Int {
        if (middlePositions.isNullOrEmpty()) return 0
        val n = middlePositions.size
        var from = max(0, min(n - 1, a))
        var to = max(0, min(n - 1, b))
        if (from > to) {
            val t = from
            from = to
            to = t
        }
        var sum = 0.0
        for (i in from until to) {
            sum += dist2d(middlePositions[i], middlePositions[i + 1])
        }
        return round(sum).toInt()
    }

    @JvmStatic
    fun compute(middlePositions: List<BlockPos>, spans: List<Records.RoadSpan>?): RangeResult {
        val n = middlePositions.size
        val isBridge = BooleanArray(n)
        val skipSegments = BooleanArray(n)
        var bridgeRanges: MutableList<IntArray> = ArrayList()

        val cfg: ModConfig = ConfigService.get()
        if (!cfg.bridgeEnabled()) {
            return RangeResult(isBridge, listOf(), skipSegments)
        }

        val minLength = max(1, cfg.bridgeMinLength())
        val mergeGap = max(1, cfg.bridgeMergeGap())
        val useBuoysInstead = cfg.bridgeUseBuoysInstead()
        val maxLenBlocks = if (useBuoysInstead) 0 else max(0, cfg.bridgeMaxLengthBlocks())

        if (spans != null) {
            val indexMap = HashMap<Long, Int>()
            for (i in 0 until n) indexMap[middlePositions[i].asLong()] = i
            for (sp in spans) {
                if (sp.type != Records.SpanType.BRIDGE) continue
                val si = indexMap[sp.start.asLong()]
                val ei = indexMap[sp.end.asLong()]
                if (si == null || ei == null) continue
                val a = max(0, min(si, ei))
                val b = min(n - 1, max(si, ei))
                bridgeRanges.add(intArrayOf(a, b))
            }
        }

        if (bridgeRanges.isNotEmpty()) {
            // 按起点排序
            bridgeRanges.sortWith(compareBy { it[0] })

            // 第一轮：合并间隔小于 mergeGap 的区间（避免频繁起伏）
            val merged = ArrayList<IntArray>()
            var cur = bridgeRanges[0]
            for (idx in 1 until bridgeRanges.size) {
                val nxt = bridgeRanges[idx]
                if (nxt[0] <= cur[1] + mergeGap) {
                    cur[1] = max(cur[1], nxt[1])
                } else {
                    merged.add(cur)
                    cur = nxt
                }
            }
            merged.add(cur)

            // 第二轮：过滤掉太短的桥梁区间（避免小水坑建桥）
            val filtered = ArrayList<IntArray>()
            for (r in merged) {
                val len = r[1] - r[0] + 1
                if (len >= minLength) {
                    filtered.add(r)
                }
            }

            // 第三轮：过滤掉过长的桥梁区间（避免跨海桥），并标记需要跳过生成的水域段
            val filteredByMaxLen = ArrayList<IntArray>()
            for (r in filtered) {
                if (maxLenBlocks <= 0) {
                    filteredByMaxLen.add(r)
                    continue
                }
                val approxLen = estimateRangeLengthBlocks(middlePositions, r[0], r[1])
                if (approxLen > maxLenBlocks) {
                    val s = max(0, r[0])
                    val e = min(n - 1, r[1])
                    for (k in s..e) {
                        skipSegments[k] = true
                    }
                } else {
                    filteredByMaxLen.add(r)
                }
            }
            bridgeRanges = filteredByMaxLen
        }

        // 根据最终的桥梁区间设置 isBridge 标记
        for (r in bridgeRanges) {
            for (k in r[0]..r[1]) {
                isBridge[k] = true
            }
        }

        return RangeResult(isBridge, bridgeRanges, skipSegments)
    }
}
